From c519d7a6e53e15e144b81d7343b1808885d782e1 Mon Sep 17 00:00:00 2001
From: Jesse Cai <jcjessecai@gmail.com>
Date: Mon, 18 Dec 2023 03:13:14 -0800
Subject: [PATCH] [sparse] update cslt to 0.5.2.1 (#115988)

Summary:

- update install_cusparselt to download 0.5.2.1 for 12.1
- add ifdef for new compute_type changes

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/115988
Approved by: https://github.com/malfet
ghstack dependencies: #115369
---
 .../ATen/native/sparse/cuda/cuSPARSELtOps.cpp | 20 +++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp b/aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp
index 701aca63a49..9091badcba3 100644
--- a/aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp
+++ b/aten/src/ATen/native/sparse/cuda/cuSPARSELtOps.cpp
@@ -137,6 +137,25 @@ at::Tensor _cslt_sparse_mm(
         compression_factor = 10;
 
         break;
+// cuSPARSELt v0.5.2 onwards changes CUSPARSE_COMPUTE_TF32, CUSPARES_COMPUT_16F to CUSPARSE_COMPUTE_32F
+#if defined(CUSPARSELT_VERSION) && CUSPARSELT_VERSION >= 502
+    case at::ScalarType::Half:
+        input_type = CUDA_R_16F;
+        output_type = CUDA_R_16F;
+        compute_type = CUSPARSE_COMPUTE_32F;
+        break;
+    case at::ScalarType::BFloat16:
+        input_type = CUDA_R_16BF;
+        output_type = CUDA_R_16BF;
+        compute_type = CUSPARSE_COMPUTE_32F;
+        break;
+    case at::ScalarType::Float:
+        input_type = CUDA_R_32F;
+        output_type = CUDA_R_32F;
+        compute_type = CUSPARSE_COMPUTE_32F;
+        break;
+// cuSPARSELt <= v0.5.2 uses CUSPARSE_COMPUTE_TF32, CUSPARES_COMPUTE_16F
+#else
     case at::ScalarType::Half:
         input_type = CUDA_R_16F;
         output_type = CUDA_R_16F;
@@ -152,6 +171,7 @@ at::Tensor _cslt_sparse_mm(
         output_type = CUDA_R_32F;
         compute_type = CUSPARSE_COMPUTE_TF32;
         break;
+#endif
     default:
         TORCH_CHECK(false, "Unsupported dtype for cuSPARSE compressed matrix multiplication.");
         break;
-- 
2.43.0

